### ISSUE TITLE
proposal: consider a `unnecessary_library_names` lint

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#3882]
Current guidance discourages named libraries (https://github.com/dart-lang/site-www/commit/dbac48c8adfb4bd93462440f54a936f0036ed5f5) and we should consider a lint to support the advice.

Related lints:

* [library_names](https://dart-lang.github.io/linter/lints/library_names.html)
* [unnecessary_library_directive](https://dart-lang.github.io/linter/lints/unnecessary_library_directive.html)

/fyi @MaryaBelanger @munificent @srawlins

**Comment by MaryaBelanger** [comment]
My take:
1. [library_names](https://dart-lang.github.io/linter/lints/library_names.html) shouldn't be a lint anymore
  (it conflicts or at least confuses "no named libraries" if we also promote a "right" way to name libraries)
2. Add a new lint called `unnecessary_library_name`


Something like this: 
```dart
const _name = 'unnecessary_library_name';

const _desc =
    'Avoid library names.';

const _details = r'''
From the [style guide](https://dart.dev/guides/language/effective-dart/style#dont-explicitly-name-libraries):

**DON'T** explicitly name libraries by attaching a name to a library directive.

**BAD:**
library my_library;

**GOOD:**
library;

''';
// ....

class UnnecessaryLibraryName extends LintRule {
  static const LintCode code = LintCode(
    _name,
    'Library directives should not have names.',
    correctionMessage: 'Try deleting the library name.',
  );
```

_(not set on `From the [style guide]...` but that's a different conversation)_

**Comment by pq** [comment]
I *think* we're in agreement so I'm bumping this to "accepted".  Chime in if you have reservations!

**Comment by bwilkerson** [comment]
I also prefer the name `unnecessary_library_name`.

But I have a question: will the lint still fire if the library has parts and at least one of the parts uses the name of the library?

I'd kind of prefer that the answer be "yes", and we can easily automate converting the part-of directive to use a URL, but it is harder for users to update manually and there are code generators that might still be using names.

I'd like to see a more complete specification before we mark this as accepted.

**Comment by natebosch** [comment]
> will the lint still fire if the library has parts and at least one of the parts uses the name of the library?

I don't think we care too much either way. We already have https://dart.dev/tools/linter-rules/use_string_in_part_of_directives in our recommended set and any project which is conformant to this one will not be impacted either way by this decision.

Whichever is easier to implement should be fine. If it's easy enough to allow library names when there is a part relying on that name I'd lean towards doing that since it might be less confusing in edge cases, but I wouldn't bother if it's significantly more effort.

**Comment by lrhn** [comment]
I'd be fine with just saying "no library names" and not checking for whether they are used.
If one really needs a library name for something, `// ignore:` works.

**Comment by pq** [comment]
It looks like we're shooting to include this in the next `package:lints` release, so I'll take a crack at implementing.

**Comment by pq** [comment]
Implemented w/ https://github.com/dart-lang/sdk/commit/71456592d0268238ac0f8a5c7d3fd8bf2c096752

Quick-fix in flight: https://dart-review.googlesource.com/c/sdk/+/358681
